### ISSUE TITLE
Autogluon imputer

### ISSUE THREAD

**Issue by obwohl** [autogluon/autogluon#3887]
Hello,

I've developed an iterative iterative_autogluon_imputer for datasets. It's roughly inspired by scikit iterative imputer but works differently.
Feel free to use or modify it. Surely it can be implemented WAY better - this is just a (working) proof of concept. 
I believe this could be a valuable addition to the AutoGluon toolkit, particularly for users dealing with datasets that have a significant amount of missing data. It seems just statistical guessing for best imputation is not a great method if there is such a powerful predictor like autogluon. I know for training with ag that the most elegant way is to let each model handle im in its way but maybe there is a need for imputation anyway.

# What It Does
The `iterative_autogluon_imputer` method iteratively imputes missing values in a dataframe by:

0. **Baseline fit and eval** fit label on original df with ag internal imputation methods.
1. **Identifying Columns with Missing Values**
2. **Iterative Imputation**: For each column with missing values, fit a model on the subset of all its non-NaN rows. Then using that predictor to predict the NaN values (just for this exact column obv.). Then combine the predictions for the NaN with the non-NaN rows.
3. **Performance Evaluation**: After each imputation, the method evaluates the impact on the label's prediction accuracy. This is done by fitting on the label with the newly imputed data (one for each feature) and comparing performance metrics.
4. **Best Improvement Selection**: It selects the imputation that most improves the label's prediction accuracy and updates the dataset accordingly.
5. **Fallback Imputation**: Optionally, if no improvement is found, the method can apply a fallback strategy (median for numerical columns and mode for non-numerical ones).
6. **Repeating the Process**: This process is repeated until all missing values are imputed or no further improvements are observed.

in short:
fit baseline: label on original df
identify columns with NaN
fit and impute all features one-by-one
for each of those: fit and eval label on the imputed data.
take the best (max) of these and compare to baseline (one might implement a significance test here in the future)
if imputed is better, make that df the new baseline (new baseline df, and metric)
elif fallback: impute the column with the fewest NaN-values just with median/mode and make this the new baseline
else: break (programm ends)

### Why it's cool
- **Flexibility**: Users can specify the time limit (per .fit) and quality settings for AutoGluon models.
- **Error Metrics**: Customizable error metric for evaluating improvements in label prediction.
- **Fallback Strategy**: Optional fallback imputation for cases where iterative imputation does not yield improvement.

### Why it may suck
- **Overfitting**:  Maybe this method yields brutal overfitting, but that's hard to predict.
- **Code structure**: I suck at coding so it's not very streamlined or pythonic. It works though if you can handle some problems. Feel free to use and optimize that code if you like.
- **Slow**: This is a very slow and thorough method. If you have f NaN features it can take up to (2f^2+2f+1)*seconds_per_fit seconds. So if you don't have time use some .fillna() you like or scikit.

### Possible improvements (besides from writing good code)
- **Fallback methods**: maybe there are better methods to revert to if there is no direct improvement than median/mode
- **always_take_max - parameter**: maybe the imputer should always take the maximum improvement disregarding if it is better than the baseline (no more median-fallback necessary then)

`def train_on_column(df, column, time=5, quality="medium_quality"):
    # Determine if the column is for a regression problem
    unique_numerical_values = pd.to_numeric(df[column], errors='coerce').nunique(dropna=True)
    PROBLEM_TYPE = 'regression' if unique_numerical_values > 10 else None
  
    print(f"fitting {column}")
    
    # Prepare training data
    df_train = df.dropna(subset=[column])
    autogluon_df = TabularDataset(df_train)
    
    # Train the predictor
    predictor = TabularPredictor(label=column, verbosity=1, problem_type=PROBLEM_TYPE).fit(
        autogluon_df, presets=quality, time_limit=time)
    
    # Evaluate the predictor
    metrics = predictor.evaluate(autogluon_df)

    return predictor, metrics
def fit_impute_single_feature(df, column, time=5, quality="medium_quality"):
    
    # fit
    non_nan_df= df.dropna(subset=[column])
    nan_df = df[pd.isna(df[column])]
    predictor, metrics = train_on_column(non_nan_df, column, time, quality)

    # impute
    predictions= predictor.predict(nan_df)
    df_imput = df.copy(deep=True)
    df_imput.loc[predictions.index, column] = predictions
    
    return df_imput

def iterative_autogluon_imputer(df, label, ERROR_METRIC= "mcc", time=5, quality="medium_quality", use_fallback_imputation=False):

    columns_with_nan = df.columns[df.isna().any()].tolist()

    # safe a copy for comparison in the end
    start_nan_columns = copy.deepcopy(columns_with_nan)

    print("bla1")
    print(set(start_nan_columns)-set(columns_with_nan))
    print("bla2")

    f=len(columns_with_nan)

    print(f"Your df has {f} NaN - columns. The imputation might take up to {time*(2*f**2+2*f+1)} seconds.")
    baseline_predictor, baseline_metric = train_on_column(df, label, time)

    while f>0:
        metrics_label_imputed={}
        predictors_label_imputed={}
        imputed_dfs={}

        for col in columns_with_nan:
            df_imput = fit_impute_single_feature(df, col)
            label_on_imputed_predictor, label_on_imputed_metric= train_on_column(df_imput, label, time, quality)
            print("#####################")
            print(f"{label} on baseline df:      {baseline_metric} ({ERROR_METRIC})")
            print(f"{label} on {col}-imputed df: {label_on_imputed_metric} ({ERROR_METRIC}")
            print("#####################\n")
            predictors_label_imputed[col]=label_on_imputed_predictor

            metrics_label_imputed[col]=label_on_imputed_metric
            
            imputed_dfs[col]=df_imput

        # Corrected code for finding the best improvement and updating the baseline
        max_metric = float('-inf')
        improver_key = None
        for col, metrics in metrics_label_imputed.items():
            if metrics.get(ERROR_METRIC, float('-inf')) > max_metric:
                max_metric = metrics[ERROR_METRIC]
                improver_key = col

        if improver_key is not None and max_metric > baseline_metric[ERROR_METRIC]:
            print("################################################")
            print(f"New imputed baseline found with {improver_key}")
            print(f"METRICS({ERROR_METRIC})\n base: {baseline_metric[ERROR_METRIC]}\n {improver_key}: {metrics_label_imputed[improver_key][ERROR_METRIC]}")
            print("################################################\n")
            baseline_predictor = predictors_label_imputed[improver_key]
            df = imputed_dfs[improver_key]
            columns_with_nan = df.columns[df.isna().any()].tolist()
            baseline_metric = metrics_label_imputed[improver_key]
            print(f"Updated baseline_metric: {baseline_metric}, type: {type(baseline_metric)}")
        
        elif use_fallback_imputation:
            # Identify the column with the smallest number of missing values
            min_nan_col = columns_with_nan[0]
            min_nan_count = df[min_nan_col].isna().sum()
            for col in columns_with_nan:
                nan_count = df[col].isna().sum()
                if nan_count < min_nan_count:
                    min_nan_col = col
                    min_nan_count = nan_count

            # Check if the column is numeric or non-numeric
            if pd.api.types.is_numeric_dtype(df[min_nan_col]):
                # Impute numeric columns with median
                df[min_nan_col].fillna(df[min_nan_col].median(), inplace=True)
            else:
                # Impute non-numeric columns with mode (most frequent value)
                mode_value = df[min_nan_col].mode().iloc[0]
                df[min_nan_col].fillna(mode_value, inplace=True)

            print("################################################")
            print(f"Imputed column {min_nan_col} with {'median' if pd.api.types.is_numeric_dtype(df[min_nan_col]) else 'mode'} due to lack of improvement in prediction.")
            print("################################################")

            # Update
            columns_with_nan = df.columns[df.isna().any()].tolist()
            baseline_predictor, baseline_metric = train_on_column(df, label, time, quality)

        else:
            print("No improvement found and fallback imputation is disabled.")
            break

        f -= 1
        print("#####################")
        print(f"{f} columns left to impute")
        print("#####################")

            
        
    print("################################################")
    print("Finished")
    return df, baseline_predictor, baseline_metric`

**Comment by ladri8** [comment]
interested. thanks

**Comment by obwohl** [comment]
https://github.com/obwohl/autogluon-imputer

**Comment by Innixma** [comment]
Cool stuff @obwohl, thanks for sharing!

It could be something interesting for us to make an official version of this, although the tricky part will be demonstrating performance improvements over the current NaN logic which mostly treats NaNs as their own values rather than imputing them.
